contracts/tgrade-valset/src/msg.rs

```diff
@@ -306,6 +204,150 @@ pub enum QueryMsg {
+    pub ratio: Decimal,
+}
+
+impl UnvalidatedDistributionContract {
+    fn validate(self, api: &dyn Api) -> Result<DistributionContract, ContractError> {
+        Ok(DistributionContract {
+            contract: api.addr_validate(&self.contract)?,
+            ratio: self.ratio,
+        })
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug, Default)]
+#[serde(transparent)]
+pub struct UnvalidatedDistributionContracts {
+    pub inner: Vec<UnvalidatedDistributionContract>,
+}
+
+impl UnvalidatedDistributionContracts {
+    /// Validates the addresses and the sum of ratios.
+    pub fn validate(self, api: &dyn Api) -> Result<Vec<DistributionContract>, ContractError> {
+        if self.sum_ratios() > Decimal::one() {
+            return Err(ContractError::InvalidRewardsRatio {});
+        }
+
+        self.inner.into_iter().map(|c| c.validate(api)).collect()
+    }
+
+    fn sum_ratios(&self) -> Decimal {
+        self.inner
+            .iter()
+            .map(|c| c.ratio)
+            .fold(Decimal::zero(), Decimal::add)
+    }
+}
+
+pub fn default_fee_percentage() -> Decimal {
+    Decimal::zero()
+}
+
+pub fn default_validators_reward_ratio() -> Decimal {
+    Decimal::one()
+}
+
+pub fn default_double_sign_slash() -> Decimal {
+    Decimal::percent(50)
+}
+
+/// Validator Metadata modeled after the Cosmos SDK staking module
+#[derive(
+    Serialize, Deserialize, Clone, Eq, PartialEq, Ord, PartialOrd, JsonSchema, Debug, Default,
+)]
+pub struct ValidatorMetadata {
+    /// The validator's name (required)
+    pub moniker: String,
+
+    /// The optional identity signature (ex. UPort or Keybase)
+    pub identity: Option<String>,
+
+    /// The validator's (optional) website
+    pub website: Option<String>,
+
+    /// The validator's (optional) security contact email
+    pub security_contact: Option<String>,
+
+    /// The validator's (optional) details
+    pub details: Option<String>,
+}
+
+pub const MIN_MONIKER_LENGTH: usize = 3;
+pub const MIN_METADATA_SIZE: usize = 1;
+pub const MAX_METADATA_SIZE: usize = 256;
+
+impl ValidatorMetadata {
+    pub fn validate(&self) -> Result<(), ContractError> {
+        if self.moniker.len() < MIN_MONIKER_LENGTH || self.moniker.len() > MAX_METADATA_SIZE {
+            return Err(ContractError::InvalidMetadata {
+                data: "moniker".to_owned(),
+                min: MIN_MONIKER_LENGTH,
+                max: MAX_METADATA_SIZE,
+            });
+        }
+        if let Some(identity) = &self.identity {
+            if identity.is_empty() || identity.len() > MAX_METADATA_SIZE {
+                return Err(ContractError::InvalidMetadata {
+                    data: "identity".to_owned(),
+                    min: MIN_METADATA_SIZE,
+                    max: MAX_METADATA_SIZE,
+                });
+            }
+        }
+        if let Some(website) = &self.website {
+            if website.is_empty() || website.len() > MAX_METADATA_SIZE {
```

Comment:
We should allow empty website

Reply:
Ah, None is okay, but Some("") not.
Fair

Let's ensure this starts with http:// or https:// (if not empty)